stylus-sdk/src/host/mod.rs

```diff
@@ -0,0 +1,364 @@
+    /// `return_data_len` will store the length of the revert data, the bytes of which can be read
+    /// via the `read_return_data` hostio. The semantics are equivalent to that of the EVM's
+    /// `[CREATE2`] opcode, which notably includes the exact address returned.
+    ///
+    /// [`Deploying Stylus Programs`]: https://docs.arbitrum.io/stylus/quickstart
+    /// [`CREATE2`]: https://www.evm.codes/#f5
+    fn create2(
+        &self,
+        code: Address,
+        endowment: U256,
+        salt: Option<B256>,
+    ) -> Result<Address, Vec<u8>>;
+}
+
+/// Provides access to storage access and mutation via host methods.
+pub trait StorageAccess {
+    /// Emits an EVM log with the given number of topics and data, the first bytes of which should
+    /// be the 32-byte-aligned topic data. The semantics are equivalent to that of the EVM's
+    /// [`LOG0`], [`LOG1`], [`LOG2`], [`LOG3`], and [`LOG4`] opcodes based on the number of topics
+    /// specified. Requesting more than `4` topics will induce a revert.
+    ///
+    /// [`LOG0`]: https://www.evm.codes/#a0
+    /// [`LOG1`]: https://www.evm.codes/#a1
+    /// [`LOG2`]: https://www.evm.codes/#a2
+    /// [`LOG3`]: https://www.evm.codes/#a3
+    /// [`LOG4`]: https://www.evm.codes/#a4
+    fn emit_log(&self, input: &[u8], num_topics: usize);
+    /// Reads a 32-byte value from permanent storage. Stylus's storage format is identical to
+    /// that of the EVM. This means that, under the hood, this hostio is accessing the 32-byte
+    /// value stored in the EVM state trie at offset `key`, which will be `0` when not previously
+    /// set. The semantics, then, are equivalent to that of the EVM's [`SLOAD`] opcode.
+    ///
+    /// Note: the Stylus VM implements storage caching. This means that repeated calls to the same key
+    /// will cost less than in the EVM.
+    ///
+    /// [`SLOAD`]: https://www.evm.codes/#54
+    fn storage_load_bytes32(&self, key: U256) -> B256;
+    /// Writes a 32-byte value to the permanent storage cache. Stylus's storage format is identical to that
+    /// of the EVM. This means that, under the hood, this hostio represents storing a 32-byte value into
+    /// the EVM state trie at offset `key`. Refunds are tabulated exactly as in the EVM. The semantics, then,
+    /// are equivalent to that of the EVM's [`SSTORE`] opcode.
+    ///
+    /// Note: because the value is cached, one must call `storage_flush_cache` to persist it.
+    ///
+    /// [`SSTORE`]: https://www.evm.codes/#55
+    fn storage_cache_bytes32(&self, key: U256, value: B256);
+    /// Persists any dirty values in the storage cache to the EVM state trie, dropping the cache entirely if requested.
+    /// Analogous to repeated invocations of [`SSTORE`].
+    ///
+    /// [`SSTORE`]: https://www.evm.codes/#55
+    fn flush_cache(&self, clear: bool);
+}
+
+/// Provides access to calling other contracts using host semantics.
+pub trait CallAccess {
```

Comment:
Note: I don't think we'll keep the call access this way, but instead keep the `call` module of the SDK and also parametrize it with the host. The way calls work in the SDK is through a builder pattern for a `Call` that provides users with a mutating call context or static call context, respectively. We probably don't need this Call trait, but I want to think about this more

Reply:
The same goes for `DeployAccess`

Reply:
The benefit of keeping it here is that it is obvious what all the host affordances are from the single trait. Perhaps the CallAccess could provide you with a CallBuilder struct? That might be ideal

Reply:
Yep we can keep it here, and then have a more ergonomic call builder that takes in a host reference on initialization, which makes calling a lot easier than using the raw host methods